src/network.rs

```diff
@@ -121,6 +123,60 @@ pub enum ReadWriteOrder {
     WriteFirst,
 }
 
+/// A wrapper around a raw `&[u8]` buffer that tracks a cursor within the buffer
+/// to allow partial fills across cancelled futures
+///
+/// Similar to `tokio::io::ReadBuf` but takes ownership of the underlying buffer to
+/// avoid coloring interfaces with lifetime parameters
+#[derive(Debug)]
+struct BufferWithCursor {
```

Comment:
I have low context here, but could we use a `std::io::Cursor` instead of defining our own struct?

Reply:
Hadn't seen this so appreciate the find. The useful methods on `io::Cursor` (e.g. `remaining_slice` and `is_empty`) are unstable [see here](https://doc.rust-lang.org/std/io/struct.Cursor.html#method.remaining_slice) and by discussion it looks like that API may change [see here](https://github.com/rust-lang/rust/pull/109174). So yeah I'd prefer to use `io::Cursor` but maybe we switch it after that interface stabilizes. I'll add a todo